### DISCUSSION TITLE
Increased memory consumption for Kubernetes executor after upgrade

### DISCUSSION THREAD

**Discussion by freddiebru** [apache/airflow#31706] [General] [answered]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

We have made upgrade of airflow version from 2.2.5 to 2.5.3. We are running everything using KubernetesExecutor and we notice an increase in memory consumption even in the most simple tasks.
Until now we have used 256mb as the baseline for request, and now we need at least 300mb.

### What you think should happen instead

We would expect that for the same code , memory consumption will remain as it was in previous version.

### How to reproduce

1. Deploy airflow version 2.2.5 with Kubernetes executor . Create simple dag with run sql queries against redshift or PostgresSQL. Run the dag several times and collect statistics of memory consumptions of pods that running in Kubernetes.
2. Do the same as in previous paragraph. but with the version 2.5.3
3. Compare average memory consumption within pods between two versions.

### Operating System

Kuberentes v1.23.17

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==2.4.0
apache-airflow-providers-cncf-kubernetes==3.0.0
apache-airflow-providers-common-sql==1.4.0
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-google==6.1.0
apache-airflow-providers-http==4.3.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-jdbc==3.2.1
apache-airflow-providers-microsoft-mssql==2.0.1
apache-airflow-providers-mysql==3.2.1
apache-airflow-providers-postgres==5.2.2
apache-airflow-providers-sqlite==3.3.1

### Deployment

Other 3rd-party Helm chart

### Deployment details

Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.17-eks-0a21954", GitCommit:"cd5c12c51b0899612375453f7a7c2e7b6563f5e9", GitTreeState:"clean", BuildDate:"2023-04-15T00:32:27Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}

### Anything else

Problem occurs consistently after upgrade 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
Where do you draw the conclusion that memory used MUST remain the same between versions? My feeling is that when new features are added, memory used might grow. But somehow your assumption is that memory used is fixed. Can you please explain where the assumption of yours is based on? 

Converted it into discussion until this point is clarified.

**Reply by freddiebru** [reply, marked as answer]
My assumption is based on fact that if the same task from previous version of airflow running (with the same code) in container/pod of Kubernetes it should consumpt more or less the same amount of memory. We have seen that memory consumption increased up to 30% , and it looks unreasonable especially if airflow running a lot of tasks it could cost a lot of money finally. Also we are not talking here about webserver or scheduler containers in Kubernetes which are directly affected from version upgrade, the memory increased in tasks running container which as I understand wrapped with new version of airflow , but I wouldn't expect that it will cause to such significant growth in memory consumption.

**Reply by potiuk** [reply]
The overhead migh come from many sources - for example versions of libraries your DAGs are using, the code run in your DAG, Top-level code that is parsed in your DAG, imports that are executed from your DAG.

Your expectations and assumptions are unfortunately wrong. This is not how airflow works. Probably you nee to revisit your knowledge and apply better deployment model for you (likely CeleryKubernetesExecutor is best for you if you care about individual task overhead).

Whenever task is run, the POD will load and start new Python interpreter, loads pretty much most of the airflow code, but also all the code you have writen and all the libraries that are used by your code are initialized. So expectations that things are not changing when new version of Airlfow is updated with new features, is wrong, The same airflow is loaded in webserver, scheduler and task and it has new features (for example in your case what you get is deferrable operators which are targeted to save you money if you convert your code to use them). Deferrable operators when you switch to them instead of sensors are showing sometimes 90% of cost saving for Airlfow users - that's why newer airflow version might be a bit bigger to accomodate for that. That - in turn - might save you money, if you just switch to using them.


Again it could be that those features are triggering for example some initial memory used during imports from the side of your DAGS, that's why it's quite impossible to run any investigation without seeing any details and analysis from your monitoring - what exactly takes more memory. If you could dig-in in your monitoring and provide some more dtails - comparing the memory used in your deployment by different components/libraries etc. that could be helpful.

What you are likely hitting also is lack of understanding that the choice of deployment you chose when you chose K8S executor is not exactly optimized for memory used. If you read at documents of K8S executor, and CeleryExecutor - the overhead is mentioned as expected in K8S executor and the way to minimise it is to use CeleryExecutor or combine the two (and use CeleryKubernetesExecutor) and route your different tasks differently.

K8S executor provide excellent isolation, but they also mean exactly what you observed, even small change in memory footprint is multiplied by the number of tasks that you run and yes - that is a lot. K8S executor shoudl really be used when the overhead of runnig a POD  is small compating to the memory/cpu footprint of the task you ar running. It's not uncommon that memory footprint of tasks that you run goes in GBs, so the 50 MB extra is not 30% but merely few percent of the overall task footprint.  If you want to optimize memory used - you should  use Celery or CeleryKubernetesExecutor, have CeleryWorkers runnig your tasks and in this case the overall memory footprint for your tasks is much smaller, becaus - unlike isolation of containers, Celery is based on process forking and process forking uses shared memory that uses "copy-on-write" memory scheme and effectively the 250MB or 300 MB of initial footprint are shared across all paralel processes that run in single worker. That also allows you to save a lot of CPU and I/O, because your Celery worker is started only once and takes tasks from its queue. Contrast that with K8S executor - every smallest task will intiialize container, allocate new memory, start new interpreter, loads all airflow code import all your DAG code and only THEN run your code. That;s a lot of overhead and if you have small tasks that you are worried about baseline of, you can probably save a lot more by properly choosing the deployment model - for example running all short tasks that do not require full isolation via Celery and all tasks that are longer and need isolation via Kubernetes queue in CaleryKubernetesExecutor.

I suggest you look at that and optimise it - because in the future the individual overhead might be even bigger, but your higher cost comes from your deployment not being designed to optimise your case.

**Comment by hussein-awala** [comment]
I agree with @potiuk 
> The overhead migh come from many sources - for example versions of libraries your DAGs are using, the code run in your DAG, Top-level code that is parsed in your DAG, imports that are executed from your DAG.

I just tested the KuberenetesExecutor with 2.2.5 and 2.5.3 to run a small python operator, and here is the result:
In the case of Airflow 2.2.5, the worker process consumed 109 Mb of memory:
![Screenshot from 2023-06-18 18-20-50](https://github.com/apache/airflow/assets/21311487/0dd414a2-97d9-4183-b4e4-302ec72735ff)
On the other hand, Airflow 2.5.3 used 155 Mb of memory
![Screenshot from 2023-06-18 18-06-15](https://github.com/apache/airflow/assets/21311487/696bb405-9c20-495a-ac40-f53b57d89a6e)
This represents an increase of 42% in my case. However, it's important to note that the difference is just 46 Mb. Considering these relatively small values, it would be more meaningful to mention the actual size rather than focusing on the percentage.


Does the memory usage difference depend solely on the Airflow version?

The memory usage difference can be attributed to the Airflow version, as it includes additional features that may result in increased memory consumption. However, it's worth noting that the impact of other library versions and their memory usage should also be taken into consideration. When comparing the sizes of Python packages between the two versions, significant differences can be observed, which can directly affect overall memory usage. Nevertheless, the crucial aspect lies in how the new versions of these libraries handle memory management and utilize resources efficiently.
Airflow 2.2.5:
```
$ docker run -i --entrypoint du apache/airflow:2.2.5 -h --max-depth=1 /home/airflow/.local/lib/python3.7/site-packages/ | sort -h
...
1.0M	/home/airflow/.local/lib/python3.7/site-packages/psutil
1.1M	/home/airflow/.local/lib/python3.7/site-packages/alembic
1.1M	/home/airflow/.local/lib/python3.7/site-packages/dns
1.2M	/home/airflow/.local/lib/python3.7/site-packages/authlib
1.2M	/home/airflow/.local/lib/python3.7/site-packages/boto3
1.2M	/home/airflow/.local/lib/python3.7/site-packages/distlib
1.2M	/home/airflow/.local/lib/python3.7/site-packages/zope
1.6M	/home/airflow/.local/lib/python3.7/site-packages/slack_sdk
1.7M	/home/airflow/.local/lib/python3.7/site-packages/celery
1.7M	/home/airflow/.local/lib/python3.7/site-packages/prompt_toolkit
1.8M	/home/airflow/.local/lib/python3.7/site-packages/flower
1.8M	/home/airflow/.local/lib/python3.7/site-packages/future
1.8M	/home/airflow/.local/lib/python3.7/site-packages/tornado
2.0M	/home/airflow/.local/lib/python3.7/site-packages/looker_sdk
2.0M	/home/airflow/.local/lib/python3.7/site-packages/yaml
2.2M	/home/airflow/.local/lib/python3.7/site-packages/elasticsearch
2.3M	/home/airflow/.local/lib/python3.7/site-packages/docutils
2.4M	/home/airflow/.local/lib/python3.7/site-packages/ldap3
2.5M	/home/airflow/.local/lib/python3.7/site-packages/setuptools
2.7M	/home/airflow/.local/lib/python3.7/site-packages/pytz
2.7M	/home/airflow/.local/lib/python3.7/site-packages/pytzdata
2.9M	/home/airflow/.local/lib/python3.7/site-packages/nacl
3.2M	/home/airflow/.local/lib/python3.7/site-packages/distributed
4.3M	/home/airflow/.local/lib/python3.7/site-packages/pygments
4.7M	/home/airflow/.local/lib/python3.7/site-packages/dask
5.3M	/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy
6.3M	/home/airflow/.local/lib/python3.7/site-packages/flask_appbuilder
6.7M	/home/airflow/.local/lib/python3.7/site-packages/pip
6.9M	/home/airflow/.local/lib/python3.7/site-packages/psycopg2_binary.libs
9.2M	/home/airflow/.local/lib/python3.7/site-packages/virtualenv
11M	/home/airflow/.local/lib/python3.7/site-packages/cryptography
11M	/home/airflow/.local/lib/python3.7/site-packages/grpc
13M	/home/airflow/.local/lib/python3.7/site-packages/kubernetes
15M	/home/airflow/.local/lib/python3.7/site-packages/lxml
19M	/home/airflow/.local/lib/python3.7/site-packages/gevent
22M	/home/airflow/.local/lib/python3.7/site-packages/numpy
23M	/home/airflow/.local/lib/python3.7/site-packages/swagger_ui_bundle
26M	/home/airflow/.local/lib/python3.7/site-packages/airflow
29M	/home/airflow/.local/lib/python3.7/site-packages/babel
33M	/home/airflow/.local/lib/python3.7/site-packages/numpy.libs
40M	/home/airflow/.local/lib/python3.7/site-packages/pandas
41M	/home/airflow/.local/lib/python3.7/site-packages/azure
41M	/home/airflow/.local/lib/python3.7/site-packages/mysql
66M	/home/airflow/.local/lib/python3.7/site-packages/botocore
72M	/home/airflow/.local/lib/python3.7/site-packages/google
85M	/home/airflow/.local/lib/python3.7/site-packages/pyarrow
745M	/home/airflow/.local/lib/python3.7/site-packages/
```
And Airflow 2.5.3:
```
$ docker run -i --entrypoint du apache/airflow:2.5.3 -h --max-depth=1 /home/airflow/.local/lib/python3.7/site-packages/ | sort -h
...
1.1M	/home/airflow/.local/lib/python3.7/site-packages/alembic
1.1M	/home/airflow/.local/lib/python3.7/site-packages/chardet
1.1M	/home/airflow/.local/lib/python3.7/site-packages/oscrypto
1.1M	/home/airflow/.local/lib/python3.7/site-packages/pkg_resources
1.1M	/home/airflow/.local/lib/python3.7/site-packages/psutil
1.1M	/home/airflow/.local/lib/python3.7/site-packages/rich
1.2M	/home/airflow/.local/lib/python3.7/site-packages/authlib
1.2M	/home/airflow/.local/lib/python3.7/site-packages/boto3
1.2M	/home/airflow/.local/lib/python3.7/site-packages/distlib
1.2M	/home/airflow/.local/lib/python3.7/site-packages/dns
1.2M	/home/airflow/.local/lib/python3.7/site-packages/zope
1.6M	/home/airflow/.local/lib/python3.7/site-packages/flower
1.6M	/home/airflow/.local/lib/python3.7/site-packages/google_ads-18.0.0.dist-info
1.6M	/home/airflow/.local/lib/python3.7/site-packages/slack_sdk
1.7M	/home/airflow/.local/lib/python3.7/site-packages/celery
1.7M	/home/airflow/.local/lib/python3.7/site-packages/prompt_toolkit
1.8M	/home/airflow/.local/lib/python3.7/site-packages/future
1.8M	/home/airflow/.local/lib/python3.7/site-packages/tornado
1.9M	/home/airflow/.local/lib/python3.7/site-packages/bcrypt
2.0M	/home/airflow/.local/lib/python3.7/site-packages/greenlet
2.0M	/home/airflow/.local/lib/python3.7/site-packages/yaml
2.1M	/home/airflow/.local/lib/python3.7/site-packages/looker_sdk
2.2M	/home/airflow/.local/lib/python3.7/site-packages/elasticsearch
2.3M	/home/airflow/.local/lib/python3.7/site-packages/docutils
2.4M	/home/airflow/.local/lib/python3.7/site-packages/ldap3
2.7M	/home/airflow/.local/lib/python3.7/site-packages/pytz
2.7M	/home/airflow/.local/lib/python3.7/site-packages/pytzdata
2.9M	/home/airflow/.local/lib/python3.7/site-packages/nacl
3.2M	/home/airflow/.local/lib/python3.7/site-packages/distributed
3.4M	/home/airflow/.local/lib/python3.7/site-packages/setuptools
3.6M	/home/airflow/.local/lib/python3.7/site-packages/aiohttp
4.5M	/home/airflow/.local/lib/python3.7/site-packages/pygments
4.7M	/home/airflow/.local/lib/python3.7/site-packages/dask
6.4M	/home/airflow/.local/lib/python3.7/site-packages/Cryptodome
6.6M	/home/airflow/.local/lib/python3.7/site-packages/flask_appbuilder
6.7M	/home/airflow/.local/lib/python3.7/site-packages/psycopg2_binary.libs
6.8M	/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy
7.9M	/home/airflow/.local/lib/python3.7/site-packages/pip
9.0M	/home/airflow/.local/lib/python3.7/site-packages/virtualenv
12M	/home/airflow/.local/lib/python3.7/site-packages/grpc
13M	/home/airflow/.local/lib/python3.7/site-packages/cryptography
13M	/home/airflow/.local/lib/python3.7/site-packages/kubernetes
13M	/home/airflow/.local/lib/python3.7/site-packages/uamqp
16M	/home/airflow/.local/lib/python3.7/site-packages/lxml
18M	/home/airflow/.local/lib/python3.7/site-packages/kubernetes_asyncio
20M	/home/airflow/.local/lib/python3.7/site-packages/gevent
22M	/home/airflow/.local/lib/python3.7/site-packages/numpy
30M	/home/airflow/.local/lib/python3.7/site-packages/mysql
33M	/home/airflow/.local/lib/python3.7/site-packages/babel
33M	/home/airflow/.local/lib/python3.7/site-packages/numpy.libs
40M	/home/airflow/.local/lib/python3.7/site-packages/pandas
46M	/home/airflow/.local/lib/python3.7/site-packages/azure
48M	/home/airflow/.local/lib/python3.7/site-packages/airflow
78M	/home/airflow/.local/lib/python3.7/site-packages/botocore
80M	/home/airflow/.local/lib/python3.7/site-packages/snowflake
113M	/home/airflow/.local/lib/python3.7/site-packages/pyarrow
125M	/home/airflow/.local/lib/python3.7/site-packages/google
960M	/home/airflow/.local/lib/python3.7/site-packages/
```

And FYI, you will have an extra small increase when you upgrade to 2.6.2:
![image](https://github.com/apache/airflow/assets/21311487/e890a683-afa4-4a69-a496-d7c05efede0b)

**Reply by freddiebru** [reply]
Hi  @potiuk @hussein-awala ,
We've recently made some necessary adjustments to make Airflow compatible with the KubernetesCeleryExecutor. However, during our testing phase, we've noticed that each individual task consumes approximately 256 MB of memory. Additionally, there's an allocation of around 166 MB for each ForkPoolWorker managed by Celery. This effectively results in each task consuming roughly 400 MB of memory. This observation challenges the assumption that Celery's use of shared memory would result in more resource-efficient task execution. What are we missing here?
Bellow is example of processes that runs in celery worker : 
USER PID %CPU %MEM 0.00MB COMMAND
root 64 6.5 1.6 263.34MB airflowtask
root 65 6.0 1.6 258.76MB airflowtask
root 70 6.4 1.6 258.69MB airflowtask
root 62 5.7 1.6 258.68MB airflowtask
root 74 6.9 1.6 258.62MB airflowtask
root 68 6.2 1.6 258.58MB airflowtask
root 80 10.4 1.6 258.53MB airflowtask
root 58 7.2 1.6 258.48MB airflowtask
root 77 9.4 1.6 258.36MB airflowtask
root 43 5.0 1.6 257.83MB airflowtask
root 93 9.5 1.6 257.27MB airflowtask
root 84 7.1 1.6 257.19MB airflowtask
root 92 12.6 1.6 256.23MB airflowtask
root 98 1.6 1.5 244.34MB airflowtask
root 89 6.8 1.4 223.77MB airflowtask
root 104 8.2 1.2 194.88MB airflowtask
root 105 12.8 1.2 191.49MB airflowtask
root 103 7.7 1.2 190.19MB airflowtask
root 109 13.8 1.2 189.53MB airflowtask
root 108 7.6 1.2 188.52MB airflowtask
root 107 7.6 1.1 187.65MB airflowtask
root 9 3.4 1.1 183.38MB [celeryd:celery@airflow-celery-worker-dev-0:MainProcess]
root 95 4.3 1.0 166.75MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-29]
root 102 0.4 1.0 166.75MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-30]
root 79 0.1 1.0 166.62MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-24]
root 76 0.2 1.0 166.59MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-23]
root 73 0.1 1.0 166.56MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-22]
root 69 0.1 1.0 166.54MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-21]
root 67 0.1 1.0 166.51MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-20]
root 66 0.1 1.0 166.49MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-19]
root 63 0.1 1.0 166.46MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-18]
root 61 0.1 1.0 166.45MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-17]
root 57 0.1 1.0 166.43MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-16]
root 56 0.1 1.0 166.42MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-15]
root 55 0.1 1.0 166.39MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-14]
root 54 0.1 1.0 166.34MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-13]
root 22 0.0 1.0 165.62MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-1]
root 32 0.0 1.0 165.61MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-11]
root 33 0.0 1.0 165.59MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-12]
root 29 0.0 1.0 165.59MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-8]
root 26 0.0 1.0 165.58MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-5]
root 31 0.0 1.0 165.57MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-10]
root 24 0.0 1.0 165.57MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-3]
root 27 0.0 1.0 165.57MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-6]
root 28 0.0 1.0 165.57MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-7]
root 25 0.0 1.0 165.55MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-4]
root 23 0.0 1.0 165.53MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-2]
root 30 0.0 1.0 165.34MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-9]
root 106 0.3 1.0 162.18MB [celeryd:celery@airflow-celery-worker-dev-0:ForkPoolWorker-31]
root 19 0.0 0.9 143.52MB gunicorn:master
root 20 0.0 0.9 143.37MB gunicorn:worker
root 21 0.0 0.9 143.37MB gunicorn:worker
root 1 0.2 0.1 27.07MB /usr/bin/python3/usr/local/bin/supervisord
root 34 0.0 0.0 3.74MB bash
root 10 0.0 0.0 3.38MB bash/usr/local/airflow/bash_scripts/celery_logs_clean/clean_logs.sh
root 114 0.0 0.0 3.12MB psaux
root 115 0.0 0.0 2.95MB awk{
root 17 0.0 0.0 0.74MB sleep1866

Thank you

**Reply by freddiebru** [reply]
Is it connected somehow to following discussion: https://github.com/apache/airflow/discussions/34341 ?
